### PR TITLE
[Attributor] Use more appropriate approach to check flat address space

### PR DESCRIPTION
**Pull request by shiltian**

### CODE REVIEW

**Comment by shiltian** [comment]
* **#108713** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/108713?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @shiltian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Transforms/IPO/Attributor.cpp [hunk 1]
```diff
@@ -3294,6 +3295,27 @@ InformationCache::getIndirectlyCallableFunctions(Attributor &A) const {
...
+    return getAssumedFlatAddressSpace();
+  auto *TTI = getAnalysisResultForFunction<TargetIRAnalysis>(*F);
+  if (!TTI)
+    return getAssumedFlatAddressSpace();
+  return TTI->getFlatAddressSpace();
+}
+
+unsigned InformationCache::getAssumedFlatAddressSpace() const {
+  if (targetIsGPU()) {
+    if (TargetTriple.isAMDGPU() || TargetTriple.isNVPTX()) {
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/Attributor.cpp:3309]: this is doing the targetIsGPU check twice, but it shouldn't be necessary anyway

File: llvm/lib/Transforms/IPO/Attributor.cpp [hunk 2]
```diff
@@ -3294,6 +3295,27 @@ InformationCache::getIndirectlyCallableFunctions(Attributor &A) const {
   return IndirectlyCallableFunctions;
 }
 
+unsigned InformationCache::getFlatAddressSpace(const Function *F) {
+  if (!F)
+    return getAssumedFlatAddressSpace();
+  auto *TTI = getAnalysisResultForFunction<TargetIRAnalysis>(*F);
+  if (!TTI)
+    return getAssumedFlatAddressSpace();
+  return TTI->getFlatAddressSpace();
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/Attributor.cpp:3304]: If you're just going to hardcode it based on the triple below, there's not much point in querying TTI

**Review by arsenm — COMMENTED** [review]
As I mentioned on the other review, this should really be a module level property from the TargetMachine or DataLayout

**Comment by shiltian** [comment]
I unstacked from https://github.com/llvm/llvm-project/pull/108786 to unblock this.

**Review by jdoerfert — APPROVED** [review]
LG

**Review by vidsinghal — APPROVED** [review]
LGTM

**Pull request merged**